packages/git/src/browser/git-quick-open-service.ts

```diff
@@ -343,6 +343,94 @@ export class GitQuickOpenService {
         });
     }
 
+    async stash(): Promise<void> {
+        const doStash = async (message: string) => {
+            if (this.repositoryProvider.selectedRepository) {
+                this.git.stash(this.repositoryProvider.selectedRepository, { message });
+            }
+        };
+        const quickOpenModel: QuickOpenModel = {
+            onType(lookFor: string, acceptor: (items: QuickOpenItem[]) => void): void {
+                const dynamicItems: QuickOpenItem[] = [];
+                const suffix = "Press 'Enter' to confirm or 'Escape' to cancel.";
+
+                if (lookFor === undefined || lookFor.length === 0) {
+                    dynamicItems.push(new SingleStringInputOpenItem(
+                        `Stash changes. ${suffix}`,
+                        () => doStash(lookFor)
+                    ));
+                } else {
+                    dynamicItems.push(new SingleStringInputOpenItem(
+                        `Stash changes with message: ${lookFor}. ${suffix}`,
+                        () => doStash(lookFor)
+                    ));
+                }
+                acceptor(dynamicItems);
+            }
+        };
+        this.quickOpenService.open(quickOpenModel, this.getOptions('Stash message', false));
+    }
+
+    protected async doStashAction(action: 'pop' | 'apply' | 'drop', text: string, getMessage?: () => Promise<string>) {
+        const repo = this.repositoryProvider.selectedRepository;
+        if (repo) {
+            const list = await this.git.stash(repo, { action: 'list' });
+            if (list) {
+                const quickOpenItems = list.map(stash => new GitQuickOpenItem<Git.StashEntry>(stash, async () => {
+                    try {
+                        await this.git.stash(repo, {
+                            action,
+                            id: stash.id
+                        });
+                        if (getMessage) {
+                            this.messageService.info(await getMessage());
+                        }
+                    } catch (error) {
+                        this.gitErrorHandler.handleError(error);
+                    }
+                }, () => stash.message));
+                this.open(quickOpenItems, text);
+            }
+        }
+    }
+
+    async applyStash(pop?: boolean): Promise<void> {
+        this.doStashAction(pop ? 'pop' : 'apply', `Select a stash to ${pop ? 'pop' : 'apply'}.`);
+    }
+
+    async dropStash(): Promise<void> {
+        this.doStashAction('drop', 'Select a stash entry to remove it from the list of stash entries.',
+            async () => {
+                if (this.repositoryProvider.selectedRepository) {
+                    const list = await this.git.stash(this.repositoryProvider.selectedRepository, { action: 'list' });
+                    const stashCount = list ? list.length : 0;
```

Comment:
We do not need this if we have the correct return type based on the arguments. See [here](
https://github.com/theia-ide/theia/pull/4427/files#r260305887).